Implement `-dump-minimization-hints` flag.

VitaNuo opened the pull request:
This PR implements a CC1 flag `-dump-deserialized-declaration-ranges`. The flag allows to specify a file path to dump ranges of deserialized declarations in `ASTReader`. Example usage:

```
clang -Xclang=-dump-deserialized-declaration-ranges=/tmp/decls -c file.cc -o file.o
```

Example output:
```
// /tmp/decls
{
  "required_ranges": [
    {
      "file": "foo.h",
      "range": [
        {
          "from": {
            "line": 26,
            "column": 1
          },
          "to": {
            "line": 27,
            "column": 77
          }
        }
      ]
    },
    {
      "file": "bar.h",
      "range": [
        {
          "from": {
            "line": 30,
            "column": 1
          },
          "to": {
            "line": 35,
            "column": 1
          }
        },
        {
          "from": {
            "line": 92,
            "column": 1
          },
          "to": {
            "line": 95,
            "column": 1
          }
        }
      ]
    }
  ]
}

```
Specifying the flag creates an instance of `DeserializedDeclsLineRangePrinter`, which dumps ranges of deserialized declarations to aid debugging and bug minimization.

Required ranges are computed from source ranges of Decls. `TranslationUnitDecl`, `LinkageSpecDecl` and `NamespaceDecl` are ignored for the sake of this PR.

Technical details:
* `DeserializedDeclsLineRangePrinter` implements `ASTConsumer` and `ASTDeserializationListener`, so that an object of `DeserializedDeclsLineRangePrinter` registers as its own listener.
* `ASTDeserializationListener` interface provides the `DeclRead` callback that we use to collect the deserialized Decls.
Printing or otherwise processing them as this point is dangerous, since that could trigger additional deserialization and crash compilation.
* The collected Decls are processed in `HandleTranslationUnit` method of `ASTConsumer`. This is a safe point, since we know that by this point all the Decls needed by the compiler frontend have been deserialized.
*  In case our processing causes further deserialization, `DeclRead` from the listener might be called again. However, at that point we don't accept any more Decls for processing.

ilya-biryukov commented on clang/include/clang/Driver/Options.td:
```diff
@@ -7968,6 +7968,10 @@ def print_dependency_directives_minimized_source : Flag<["-"],
   HelpText<"Print the output of the dependency directives source minimizer">;
 }
+def print_deserialized_declarations : Joined<["-"],
+  "print-deserialized-declarations-path=">,
```
I wonder about the naming here... Maybe `-dump-deserialized-declaration-ranges`?
It's long, but:
- captures the meaning better (i.e. talks about ranges, not declarations)
- says dump like the `-dump-deserialized-decls`.

VitaNuo replied:
Sure, sounds good.

ilya-biryukov commented on clang/include/clang/Driver/Options.td:
```diff
@@ -7968,6 +7968,10 @@ def print_dependency_directives_minimized_source : Flag<["-"],
+def print_deserialized_declarations : Joined<["-"],
+  "print-deserialized-declarations-path=">,
+  HelpText<"Print traces of deserialized declarations to aid debugging and minimization">,
+  MarshallingInfoString<FrontendOpts<"PrintDeserializedDeclarationsPath">>;
```
I suggest to make this flag a `CC1Option` so that `clang -print-deserialized-declarations-path` does not work ,but `clang -Xclang=-print-deserialized-declarations-path` is allowed to produce this.

These options are considered internal to the compiler and Clang does not promise any compatibility for those across versions. I think this is what we want to have at this point as more experiments are planned.

VitaNuo replied:
AFAICT it's already a CC1 option.

ilya-biryukov replied:
Makes sense, it's in the list of options which are CC1. I got confused, rarely look at this file. Thanks for clarying.

ilya-biryukov commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -121,6 +252,19 @@ FrontendAction::CreateWrappedASTConsumer(CompilerInstance &CI,
+  llvm::StringRef PrintDeserializedDeclarationsPath = CI.getFrontendOpts().PrintDeserializedDeclarationsPath;
+  if (!PrintDeserializedDeclarationsPath.empty()) {
+    std::error_code ErrorCode;
+    auto* FileStream = new llvm::raw_fd_ostream(PrintDeserializedDeclarationsPath, ErrorCode, llvm::sys::fs::OF_None);
```
Use `make_unique<raw_fd_ostream>(...)` here to avoid plain new.

VitaNuo replied:
Makes sense, thank you.

ilya-biryukov commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -139,16 +283,17 @@ FrontendAction::CreateWrappedASTConsumer(CompilerInstance &CI,
   // If there are no registered plugins we don't need to wrap the consumer
-  if (FrontendPluginRegistry::begin() == FrontendPluginRegistry::end())
-    return Consumer;
+  if (FrontendPluginRegistry::begin() == FrontendPluginRegistry::end()) {
```
`if (!Consumers.empty())` seems more appropriate here. One way to look at this option is that we add an "implicit" plugin.

VitaNuo replied:
But what happens then if `FrontendPluginRegistry` isn't empty? We would skip extra plugins, which seems wrong.

ilya-biryukov replied:
I have also moved the code around in my head, but never actually spelled it out, sorry.
My full suggestion would be:
- move this if to the end of the function after the plugin consumer registrationg is finished.
- right before the final line, do this:

```cpp
assert(Consumers.size() >= 1 && "should have added the main consumer");
if (Consumers.size() == 1)
  return std::move(Consumers.front());
return std::make_unique<MultiplexConsumer>(std::move(Consumers));
```

If the only way to produce many consumers is compiler plugins, this shortcut makes sense. Now that we also have this custom logic, I feel it adds another condition that one needs to track and just makes things more error-prone.

ilya-biryukov replied:
Also, we still don't want to create a `MultiplexConsumer` if our flag wasn't specified, right?
So I think we need to update this code anyhow.

VitaNuo replied:
Thanks for the detailed explanation, addressed that.

ilya-biryukov commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -121,6 +252,19 @@ FrontendAction::CreateWrappedASTConsumer(CompilerInstance &CI,
+      DeserializedDeclsLineRangePrinter* Printer = new DeserializedDeclsLineRangePrinter(CI.getSourceManager(), std::unique_ptr<llvm::raw_fd_ostream>(FileStream));
+      Consumers.push_back(std::unique_ptr<ASTConsumer>(Printer));
+    } else {
+      llvm::errs() << "Could not open file with path: " << PrintDeserializedDeclarationsPath << ", error: " << ErrorCode.message() << "\n";
```
NIT: maybe add a more descriptive error message, e.g. `Failed to create output file for -<flag>`?

VitaNuo replied:
Sure.

ilya-biryukov commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +51,135 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
+class DeserializedDeclsLineRangePrinter : public DelegatingDeserializationListener, public ASTConsumer {
+public:
+  explicit DeserializedDeclsLineRangePrinter(SourceManager &SM, std::unique_ptr<llvm::raw_fd_ostream> OS)
+      : DelegatingDeserializationListener(nullptr, false), SM(SM), OS(std::move(OS)) {}
```
I am not sure we can avoid passing `Previous` here and there does not seem to be any code that registers the `DeserializationListener` just yet.

VitaNuo replied:
The chain of delegating listeners isn't available at the point where we create and register this listener ([this deserial listener](https://github.com/VitaNuo/llvm-project/blob/main/clang/lib/Frontend/FrontendAction.cpp#L977) is passed through a chain in `BeginSourceFile`, whereas our new listener is created and registered in `CreateWrappedASTConsumer`).

However, it seems to me that we can switch inheriting from `DelegatingDeserializationListener` to parent `ASTDeserializationListener`. We create our own `ASTConsumer`, for which `DeserializedDeclsLineRangePrinter` is the only deserialization listener, which IMU means we don't need any chaining. On the contrary, the chain of consumers in `BeginSourceFile` is attached to a single ASTConsumer, so chaining is inevitable.

ilya-biryukov replied:
SG! I have missed the fact the `ASTConsumer` can have a deserialization listener attached to it and there is logic in Clang to handle that. The approach SG.

ilya-biryukov commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +51,135 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
 
 namespace {
 
+/// Dumps deserialized declarations.
```
NIT: remove this comment as the class name already makes it clear what it's doing.

If you want, you might use a comment to explain other non-trivial details, though, e.g. how the class must be used or why it implements certain interfaces and when the output happens.

VitaNuo replied:
> If you want, you might use a comment to explain other non-trivial details, though, e.g. how the class must be used or why it implements certain interfaces and when the output happens.

This is a good idea. I have written an extensive comment, PTAL.

ilya-biryukov replied:
I still see the old version. Did the push succeed?

VitaNuo replied:
Yes, sorry, I responded to the comment before pushing the actual change.

ilya-biryukov reviewed (changes requested):
Thanks, this looks like a good start! I have left a few comments and also feel there are a few things missing:
- basic tests
- the code registering deserialization listener
- suggestion: a commit message could benefit from a description of how we are going to use this flag (heuristics for minimization tools), describe the output format and mention that the output stability is not guaranteed (it's also another reason why to put it behind a `-cc1` flag)

ilya-biryukov commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +51,150 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
+
+  void printJson(const std::vector<RequiredRanges> &Result) {
+    *OS << "{\n";
+    *OS << "  \"required_ranges\": [\n";
```
NIT: using raw string literals would should improve the readability here:

```cpp
OS << R"(  "required_ranges": [)" << '\n';
```

VitaNuo replied:
Done.

ilya-biryukov reviewed (commented):
Thanks, the implementation mostly looks good. I've left a lot of comments, but they are mostly NITs. One major thing that we need to change is the way we test this, see the comment about relying on `jq`.

Another thought that crossed my mind is that we are actually heavily optimizing for minimization, but use a name that creates an impression we might care about something else. I don't have great ideas there, but maybe `-dump-minimization-hints`  would be appropriate? Any thoughts on this?

It would be great to link to the code consuming those hints in `cvise` too to add some context on why we are doing this.

ilya-biryukov commented:
Could you also update the PR description and change `[WIP]` to `[Clang]` in the title so that we don't accidentally forget this before comitting?

VitaNuo commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -121,6 +267,26 @@ FrontendAction::CreateWrappedASTConsumer(CompilerInstance &CI,
   if (!Consumer)
     return nullptr;
 
+  std::vector<std::unique_ptr<ASTConsumer>> Consumers;
```
The constructor of `MultiplexConsumer` won't take a `SmallVector`.

VitaNuo replied:
Done.

VitaNuo committed:
This commit implements a CC1 flag `-dump-deserialized-declaration-ranges`. The flag allows to specify a file path to dump ranges of deserialized declarations in `ASTReader`. Example usage:

```
clang -Xclang=-dump-deserialized-declaration-ranges=/tmp/decls -c file.cc -o file.o
```

Example output:
```
// /tmp/decls
{
  "required_ranges": [
    {
      "file": "foo.h",
      "range": [
        {
          "from": {
            "line": 26,
            "column": 1
          },
          "to": {
            "line": 27,
            "column": 77
          }
        }
      ]
    },
    {
      "file": "bar.h",
      "range": [
        {
          "from": {
            "line": 30,
            "column": 1
          },
          "to": {
            "line": 35,
            "column": 1
          }
        },
        {
          "from": {
            "line": 92,
            "column": 1
          },
          "to": {
            "line": 95,
            "column": 1
          }
        }
      ]
    }
  ]
}

```
Specifying the flag creates an instance of `DeserializedDeclsLineRangePrinter`, which dumps ranges of deserialized declarations to aid debugging and bug minimization.

Required ranges are computed from source ranges of Decls. `TranslationUnitDecl`, `LinkageSpecDecl` and `NamespaceDecl` are ignored for the sake of this PR.

Technical details:
* `DeserializedDeclsLineRangePrinter` implements `ASTConsumer` and `ASTDeserializationListener`, so that an object of `DeserializedDeclsLineRangePrinter` registers as its own listener.
* `ASTDeserializationListener` interface provides the `DeclRead` callback that we use to collect the deserialized Decls.
Printing or otherwise processing them as this point is dangerous, since that could trigger additional deserialization and crash compilation.
* The collected Decls are processed in `HandleTranslationUnit` method of `ASTConsumer`. This is a safe point, since we know that by this point all the Decls needed by the compiler frontend have been deserialized.
*  In case our processing causes further deserialization, `DeclRead` from the listener might be called again. However, at that point we don't accept any more Decls for processing.

VitaNuo commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +51,150 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
+
+  void DeclRead(GlobalDeclID ID, const Decl *D) override {
+    if (!IsCollectingDecls) {
+      return;
```
Done.

VitaNuo commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +51,150 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
+    if (!IsCollectingDecls) {
+      return;
+    }
+    if (!D || isa<TranslationUnitDecl>(D) || isa<LinkageSpecDecl>(D) ||
```
Done.

VitaNuo commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +51,150 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
+  using Position = std::pair<unsigned, unsigned>;
+  struct RequiredRanges {
+    StringRef Filename;
+    std::vector<std::pair<Position, Position>> FromTo;
```
Done.

VitaNuo commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +51,150 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
+
+      auto *F = SM.getFileEntryForID(SM.getFileID(R.getBegin()));
+      if (F != SM.getFileEntryForID(SM.getFileID(R.getEnd())))
+        continue;
```
Done.

VitaNuo commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +51,150 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
+                              SM.getSpellingColumnNumber(R.getEnd())}});
+    }
+
+    std::vector<RequiredRanges> Result;
```
Done.

VitaNuo commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +51,150 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
+      }
+      *OS << "      ]\n    }";
+      if (i < Result.size() - 1) {
+        *OS << ",";
```
Done.

VitaNuo commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +51,150 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
+    std::vector<std::pair<Position, Position>> FromTo;
+  };
+  void HandleTranslationUnit(ASTContext &Context) override {
+    IsCollectingDecls = false;
```
Done.

VitaNuo commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +51,150 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
+/// compiler frontend have been deserialized. In case our processing causes
+/// further deserialization, DeclRead from the listener might be called again.
+/// However, at that point we don't accept any more Decls for processing.
+class DeserializedDeclsLineRangePrinter : public ASTDeserializationListener,
```
Done.

VitaNuo commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +51,150 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
+    if (!D || isa<TranslationUnitDecl>(D) || isa<LinkageSpecDecl>(D) ||
+        isa<NamespaceDecl>(D))
+      return;
+    if (auto *DC = D->getDeclContext(); !DC || !DC->isFileContext())
```
Done.

VitaNuo commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +51,150 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
+    ASTDeserializationListener::DeclRead(ID, D);
+  }
+
+  using Position = std::pair<unsigned, unsigned>;
```
This is outdated now, after switching to struct.

VitaNuo commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +51,150 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
+  };
+  void HandleTranslationUnit(ASTContext &Context) override {
+    IsCollectingDecls = false;
+    std::vector<const Decl *> Decls = std::move(PendingDecls);
```
Done.

VitaNuo commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +51,150 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
+    // they do not break things.
+    struct FileData {
+      std::vector<std::pair<Position, Position>> FromTo;
+      std::vector<std::pair<unsigned, unsigned>> Columns;
```
Done.

VitaNuo commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +51,150 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
+      auto &Data = FileToLines[F];
+      if (!Data.Ref)
+        Data.Ref = SM.getFileEntryRefForID(SM.getFileID(R.getBegin()));
+      Data.FromTo.push_back({{SM.getSpellingLineNumber(R.getBegin()),
```
Done.

VitaNuo commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +51,150 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
+
+      llvm::sort(FromTo);
+
+      std::vector<std::pair<Position, Position>> MergedLines;
```
Done.

VitaNuo commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +51,150 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
+    printJson(Result);
+  }
+
+  void printJson(const std::vector<RequiredRanges> &Result) {
```
Done.

VitaNuo commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +51,150 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
+  void printJson(const std::vector<RequiredRanges> &Result) {
+    *OS << "{\n";
+    *OS << "  \"required_ranges\": [\n";
+    for (size_t i = 0; i < Result.size(); ++i) {
```
Done.

VitaNuo commented:
> Thanks, the implementation mostly looks good. I've left a lot of comments, but they are mostly NITs. One major thing that we need to change is the way we test this, see the comment about relying on jq.

Sure, changed to `FileCheck`.

> Another thought that crossed my mind is that we are actually heavily optimizing for minimization, but use a name that creates an impression we might care about something else. I don't have great ideas there, but maybe -dump-minimization-hints would be appropriate? Any thoughts on this?

TBH I would prefer not to do that. Although we're doing this for minimization, there's nothing particular about the deserialized ranges or the JSON format that we produce that limits its use to minimization. If someone else find this useful, I would only welcome that. So in this sense, I'd prefer not to have a very specific name.

> It would be great to link to the code consuming those hints in cvise too to add some context on why we are doing this.
We should do this as a followup, I think. The C-Vise code is undergoing too much change to link to a place that is likely to be outdated very soon.

VitaNuo committed:
Fix test on Windows

VitaNuo committed:
Remove old test.

VitaNuo committed:
Handle source ranges where `isTokenRange() == true` (the end of this range specifies the start of the last token). In this case, compute the source location just past the end of the token at this source location. Fix the test. The end column is not exclusive.

VitaNuo committed:
Relex the token past the end location of the last token in the source range. If it's a semicolon, advance the location by one token.

VitaNuo committed:
Fix formatting.

usx95 commented:
> > Another thought that crossed my mind is that we are actually heavily optimizing for minimization, but use a name that creates an impression we might care about something else. I don't have great ideas there, but maybe -dump-minimization-hints would be appropriate? Any thoughts on this?
> 
> TBH I would prefer not to do that. Although we're doing this for minimization, there's nothing particular about the deserialized ranges or the JSON format that we produce that limits its use to minimization. If someone else find this useful, I would only welcome that. So in this sense, I'd prefer not to have a very specific name.

I am in favor of using `-dump-minimization-hints` to scope it down and avoid Hyrum's law. If we want to be more descriptive, we could have this as a `CommaJoined` flag, eg. `-dump-minimization-hints=deserialized-decl-ranges,some-other-future-hint,all`.

VitaNuo committed:
Simplify the code by using `Lexer::getAsCharRange` to advance to the end location of the end token in the range.

VitaNuo committed:
Rename the flag to `dump-minimization-hints` to make it more specific.

VitaNuo committed:
Rename and fix the test.

emaxx-google commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +54,185 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
+      std::vector<std::pair<Position, Position>> FromTo;
+      OptionalFileEntryRef Ref;
+    };
+    llvm::DenseMap<const FileEntry *, FileData> FileToLines;
```
nit: `FileToRanges` or something like that

VitaNuo replied:
Sure.

emaxx-google commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +54,185 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
+  std::unique_ptr<llvm::raw_ostream> OS;
+
+  void printJson(llvm::ArrayRef<RequiredRanges> Result) {
+    *OS << "{\n";
```
A single huge JSON object is hard to work with efficiently - typically this results loading it all into memory at once (unless some convoluted methods are used). WDYT about switching to JSON Lines, e.g., with each line containing an object describing a single source file?

VitaNuo replied:
I've run some tests, and the maximum I'm getting is 4.4Mb of JSON for a very large compilation. I think this should be palatable. For a smaller compilation, it's often <1Mb.

emaxx-google replied:
Acknowledged - this size is not too bad to deal with.

VitaNuo committed:
Address review comments.

VitaNuo commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -121,6 +305,25 @@ FrontendAction::CreateWrappedASTConsumer(CompilerInstance &CI,
+          CI.getSourceManager(), std::move(FileStream)));
+    } else {
+      llvm::errs() << "Failed to create output file for "
+                      "-dump-deserialized-declaration-ranges flag, file path: "
```
Thanks!

VitaNuo commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +54,185 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
+/// compiler frontend have been deserialized. In case our processing causes
+/// further deserialization, DeclRead from the listener might be called again.
+/// However, at that point we don't accept any more Decls for processing.
+class DeserializedDeclsLineRangePrinter : public ASTConsumer,
```
I think `...SourceRangePrinter` sounds good.

VitaNuo commented on clang/lib/Frontend/FrontendAction.cpp:
```diff
@@ -49,6 +54,185 @@ LLVM_INSTANTIATE_REGISTRY(FrontendPluginRegistry)
+                                   const CharSourceRange &Range,
+                                   const LangOptions &LangOpts) {
+      // For token ranges, compute end location for end character of the range.
+      // The end location of returned range is exclusive.
```
It's just offset arithmetic, so it will always point at column N+1, no matter if it's the end of the line or not. Newline isn't a token for the sake of the compiler, so it would be harder to figure out if we are at the line end or not (and not necessary, if I understand you correctly).

VitaNuo replied:
Sure, I have modified the comment to say that column number (rather than end location) is exclusive.

VitaNuo committed:
Update the comment to explciitly mention that the column number of the end position is exclusive.
